### PR TITLE
All: Line sections on ptref

### PR DESCRIPTION
**Pull request by antoine-de**
this PR is part of the work on how to display the [line sections](https://github.com/CanalTP/navitia/blob/dev/documentation/rfc/line_sections.md)

For the moment we display the line section impact as before (as a line impact), on the impacted stop_points and impacted vehicle_journeys

### CODE REVIEW

File: source/ed/build_helper.cpp [hunk 1]
```diff
@@ -404,17 +403,29 @@ Impacter& Impacter::severity(const std::string& uri) {
 }
 
 Impacter& Impacter::on(nt::Type_e type, const std::string& uri) {
-    impact->informed_entities.push_back(dis::make_pt_obj(type, uri, *b.data->pt_data, impact));
+    link_informed_entity(dis::make_pt_obj(type, uri, *b.data->pt_data), impact, b.data->meta->production_date, get_disruption().rt_level);
```
Comment (TeXitoi) [inline source/ed/build_helper.cpp]: long line

**Comment by pbougue** [comment]
Jenkins, test this please

**Review by kinnou02 — APPROVED** [review]

File: source/jormungandr/tests/line_sections_tests.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        assert self.has_disruption(ObjGetter('stop_points', 'D_1'), 'line_section_on_line_1')
+        assert self.has_disruption(ObjGetter('stop_points', 'E_1'), 'line_section_on_line_1')
+        assert not self.has_disruption(ObjGetter('stop_points', 'F_1'), 'line_section_on_line_1')
+
+    def test_on_vehicle_journeys(self):
+        """
+        the line section disruption should be linked to the impacted vehicle journeys
+        """
+        assert self.has_disruption(ObjGetter('vehicle_journeys', 'vj:1:1'), 'line_section_on_line_1')
+        assert not self.has_disruption(ObjGetter('vehicle_journeys', 'vj:1:2'), 'line_section_on_line_1')
```
Comment (TeXitoi) [inline source/jormungandr/tests/line_sections_tests.py:68]: No journey test?

File: source/type/message.cpp [hunk 3]
```diff
@@ -40,6 +40,115 @@ namespace bg = boost::gregorian;
...
+
+            for (const auto& st : impacted_stoptimes) {
+                auto* sp = st.stop_point;
+                if (impacted_stop_points.insert(sp).second) {
+                    // it's the first time we see this stoppoint, we add the impact to it
+                    sp->add_impact(impact);
+                }
+            }
+        }
+
```
Comment (TeXitoi) [inline source/type/message.cpp]: useless empty line

File: source/type/message.h [hunk 4]
```diff
@@ -260,15 +251,29 @@ struct Impact {
...
+    boost::iterator_range<std::vector<PtObj>::const_iterator> informed_entities() const {
+        return boost::make_iterator_range(_informed_entities.begin(), _informed_entities.end());
+    }
+    boost::iterator_range<std::vector<PtObj>::iterator> mut_informed_entities() {
+        return boost::make_iterator_range(_informed_entities.begin(), _informed_entities.end());
+    }
+
+    // add the ptobj to the enformed entities and make all the needed backref
+    // Note: it's a friend function because we need the shared_ptr to the impact
+    friend void link_informed_entity(PtObj ptobj, boost::shared_ptr<Impact>& impact, const boost::gregorian::date_period&, type::RTLevel);
```
Comment (TeXitoi) [inline source/type/message.h]: Maybe a static method is more appropriate than a friend function?

**Review by TeXitoi — CHANGES_REQUESTED** [review]
:+1:

File: source/jormungandr/tests/line_sections_tests.py (hunk 2 above)
Comment (antoine-de) [reply source/jormungandr/tests/line_sections_tests.py:68]: nop they are for the next PR :wink:

**Review by TeXitoi — APPROVED** [review]

**Pull request merged**